src/inetstack/protocols/layer2/mod.rs

```diff
@@ -77,23 +77,29 @@ impl SharedLayer2Endpoint {
     }
 
     pub fn transmit_arp_packet(&mut self, remote_link_addr: MacAddress, pkt: DemiBuffer) -> Result<(), Fail> {
-        self.transmit(remote_link_addr, EtherType2::Arp, pkt)
+        let mut pkts: ArrayVec<DemiBuffer, MAX_BATCH_SIZE_NUM_PACKETS> = ArrayVec::new();
+        pkts.push(pkt);
+        self.transmit(remote_link_addr, EtherType2::Arp, pkts)
     }
 
     pub fn transmit_ipv4_packet(&mut self, remote_link_addr: MacAddress, pkt: DemiBuffer) -> Result<(), Fail> {
-        self.transmit(remote_link_addr, EtherType2::Ipv4, pkt)
+        let mut pkts: ArrayVec<DemiBuffer, MAX_BATCH_SIZE_NUM_PACKETS> = ArrayVec::new();
+        pkts.push(pkt);
+        self.transmit(remote_link_addr, EtherType2::Ipv4, pkts)
     }
 
     fn transmit(
         &mut self,
         remote_link_addr: MacAddress,
         eth2_type: EtherType2,
-        mut pkt: DemiBuffer,
+        mut pkts: ArrayVec<DemiBuffer, MAX_BATCH_SIZE_NUM_PACKETS>,
     ) -> Result<(), Fail> {
         let eth2_header: Ethernet2Header = Ethernet2Header::new(remote_link_addr, self.local_link_addr, eth2_type);
         debug!("L2 OUTGOING {:?}", eth2_header);
-        eth2_header.serialize_and_attach(&mut pkt);
-        self.layer1_endpoint.transmit(pkt)
+        for mut pkt in pkts.iter_mut() {
+            eth2_header.serialize_and_attach(&mut pkt);
```

Comment:
`pkt` is already a muatble reference, so there is no need to take a reference again. You can just use the code as below.

```suggestion
            eth2_header.serialize_and_attach(pkt);
```